src/ring_theory/dedekind_domain/integer_unit.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2022 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+
+import ring_theory.dedekind_domain.adic_valuation
+
+/-!
+# `S`-integers and `S`-units of fraction fields of Dedekind domains
+
+This file defines the ring of `S`-integers and the group of `S`-units of the field of fractions of a
+Dedekind domain, to be specialised to the case of a number field or a function field separately.
+
+TODO: notation for `S`-integers and `S`-units, proof that `S`-integers is the intersection of
+valuation rings, proof that `S`-units is the kernel of a map to a product, proof that `∅`-integers
+is the usual ring of integers, finite generation of `S`-units and Dirichlet's `S`-unit theorem.
+
+## Main definitions
+
+* `is_dedekind_domain.S_integer_ring`: the ring of `S`-integers.
+* `is_dedekind_domain.S_unit_group`: the group of `S`-units.
+
+## Main statements
+
+* `is_dedekind_domain.S_integer_unit`: the units of the `S`-integers are the `S`-units.
+
+## References
+
+* https://encyclopediaofmath.org/wiki/S-integer
+* https://en.wikipedia.org/wiki/S-unit
+
+## Tags
+
+S-integer, S-unit
+-/
+
+open_locale non_zero_divisors
+
+namespace is_dedekind_domain
+
+noncomputable theory
+
+universes u v
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R] (K : Type v) [field K]
+  [algebra R K] [is_fraction_ring R K] (S : set $ height_one_spectrum R)
+
+/-- The subring of `S`-integers of `K`. -/
+def S_integer_ring : subring K :=
```

Comment:
We rarely want to add a `def` without some API for it. In this case the minimum we should have are `simp` lemmas available from `simps`:
```suggestion
@[simps] def S_integer_ring : subring K :=
```
and likewise the other two `def`s below.